include/seahorn/Expr/HexDump.hh

```diff
@@ -0,0 +1,78 @@
+
+#pragma once
+#include "seahorn/Expr/ExprOpCore.hh"
+#include "llvm/Support/raw_ostream.h"
+
+namespace expr {
+
+namespace hexDump {
+/**
+ * A Pair is an individual cell in the hex dump. It stores the index and value
+ * expressions
+ *
+ * \note bvnums and Uints are stored as MPZ expressions
+ */
+class Pair {
+  std::pair<Expr, Expr> m_pair; // <index, value>
+  bool m_isRepeated = false;
+
+public:
+  Pair(Expr idx, Expr value, bool isRepeated = false);
+
+  bool isSortable() const;
+
+  void print(std::ostream &OS, std::pair<unsigned, unsigned> widths,
+             bool includeAscii) const;
+
+  mpz_class getIdxNum() const;
+
+  Expr getIdxExpr() const;
+
+  Expr getValueExpr() const;
+
+  bool isRepeated() const;
+
+  /// based off of the index
+  bool operator<(Pair const &pair) const;
+
+  bool operator==(Pair const &pair) const;
+
+  /// prints with ascii and no set width
+  friend std::ostream &operator<<(std::ostream &OS, Pair const &pair);
+};
+
+using IteratorPair = std::pair<std::vector<Pair>::const_iterator,
+                               std::vector<Pair>::const_iterator>;
+
+/**
+ * For MK_STRUCT expressions, splits every child into a separate hex dump
+ *
+ * Supports nested ITE expressions, arrays (store expressions) and finite maps
+ * (set expressions)
+ *
+ * If the given expression is not one of the supported expressions above, it
+ * will search its children for the first supported expression and use that.
+ */
+class HexDump {
+  class Impl;
+  Impl *m_impl;
+
+public:
+  HexDump(Expr exp, unsigned addressesPerWord = 1);
+  ~HexDump();
+
+  /// return iterators to the first hex dump
+  std::vector<Pair>::const_iterator beginFirst() const;
```

Comment:
With structs I made it so that there are multiple, separate hex dumps. When I made this I was worried that begin () and end() might now get the purpose of this across  - which  was just an iterator to the first hex dump

Reply:
Maybe I could use begin() and just make sure to comment it properly

Reply:
better comment sound good. It is also reasonable to not support hex dump of structs (if they are handled independently) and expect the client to apply hex dump on each element of the struct.

Reply:
Alright I'll improve the comment. The client shouldn't have to do anything extra with structs.  Calling print will print all of the hex dumps (without the client having to do anything different). The only difference is that you can get a vector of ranges/iterators

Reply:
I understand this part. You have decided to make things easier for the client by supporting not only array and ite expressions, but also supporting struct expressions. This makes it easier for the client, but maybe makes the interface more complex than necessary.

Handling a struct directly is good because all elements of the struct can be handled together. In our context, the secondary elements are annotations on the first element, so the hex dumps of them can be somehow combined together. While this is not done now, the fact that interface provides that possibility is nice.

On the other hand, the interpretation of structs is very specific to how we use them. Therefore, printing multiple hex dumps overlayed together might be better done in a separate module that is aware of the particular use of structs.

Reply:
Ok makes sense. I'll add in a separate module for that